### DISCUSSION TITLE
arduino config parser and formatter

### DISCUSSION THREAD

**Discussion by lerouxb** [SpenceKonde/ATTinyCore#513] [Show and tell]
As discussed here https://github.com/SpenceKonde/ATTinyCore/issues/510

I got a basic parser and formatter for arduino config files going as specified here https://arduino.github.io/arduino-cli/latest/platform-specification/

(Although admittedly I only skimmed that for now and I'm mostly going with how boards.txt, programmers.txt, platform.txt files that are actually out there in the wild works.)

The goal is so that this pseudo-code will work:

```
text = readFile('boards.txt');
lines = text.splitIntoLines();
parsedLines = lines.map(parseLine); // array/list of each line parsed out and made sense of
result = linesToObject(parsedLines);

// result is now a sane JSON-like simple object that you can walk, inspect, validate, modify and so on

// and here comes the tricky part:
formatted = formatResult(result);

// SOMEHOW
expect(formatted).to.equal(text)
```

That last line is where I'm hitting some minor snags. Not one boards.txt file's lines are in a predictable order. So I have to keep the order, somehow reconstruct it yet also somehow figure out how to deal with new things you might have added after you parsed the file.

I don't think the result I got when formatting things out again will be any different since it shouldn't matter if the foo.bar.baz=whatever lines' order changed, but I'd still feel much better about it if the order somehow got preserved.

---

Whats the use of all of this again? To make it easy to work with the .txt files in a high level language. ie. You'd be able to script changes, check if you missed something and so on.

**Comment by SpenceKonde** [comment]
Making the line order in the output match that of the input is not a requirement.. Making them *not* match is! The current order is what happened by chance as I modified it over time, via a combination of hand editing and regex substitution patterns, and is part of the problem. 

What I did on my newer cores (where I was being careful from the start, and the options are very uniform, and which I think is a lot better, is this:

```

##########################################################################
#      Tools Submenu Names
##########################################################################

menu.chip=Chip
menu.clock=Clock
menu.clocko=Clock (burn bootloader req'd, or see docs)
menu.millis=millis()/micros()
menu.uartvoltage=Voltage for UART baud (ignored for 2k/4k parts)
menu.serialevent=Support SerialEvent
menu.startuptime=Startup Time
menu.startuptimeopti=Startup Time (burn bootloader req'd)
menu.bodvoltage=BOD level if enabled (burn bootloader req'd)
menu.bodmode=BOD Mode when active/sleeping (burn bootloader req'd)
menu.eesave=Save EEPROM (burn bootloader req'd)
menu.bootloaderuart=UART for Bootloader (burn bootloader req'd)
menu.resetpin=UPDI Pin, READ DOCS or pick UPDI! (burn bootloader req'd)
menu.resetpinopti=UPDI/reset pin (burn bootloader req'd)

# The *opti menus are for Optiboot board definitions;
# they function identically, just differ in the
# information shown in the submenu name.

##########################################################################
#      Board Names
##########################################################################

atxy7.name=24-pin tinyAVR 0/1/2-series (ATtinyxxy7)
atxy6.name=20-pin tinyAVR 0/1/2-series (ATtinyxxy6)
atxy4.name=14-pin tinyAVR 0/1/2-series (ATtinyxxy4)
atxy2.name=8-pin ATtiny412/402/212/202
atxy7o.name=24-pin tinyAVR 0/1/2-series w/Optiboot (ATtinyxxy7)
atxy6o.name=20-pin tinyAVR 0/1/2-series w/Optiboot (ATtinyxxy6)
atxy4o.name=14-pin tinyAVR 0/1/2-series w/Optiboot (ATtinyxxy4)
atxy2o.name=8-pin ATtiny412/402/212/202 w/Optiboot
Xplained416.name=Attiny416 Xplained Nano (Optiboot)
Xplained817.name=Attiny817 Xplained Mini (Optiboot)

##########################################################################
#      ATtiny3227/3217/1627/1617/1607/827/817/807/427/417
##########################################################################


```
That is:
```

##########################################################################
#      Tools Submenu Names
##########################################################################

Each of the top-level submenu names, in the order that they were in originally. 

##########################################################################
#      Board Names
##########################################################################

(each boardname.name, in the same order that they'll show up - was about to say alphabetical, then I remembered that I did organize the boards in a specific order. And then checked it and realized the order I had them in was not what I wanted!), lol. See what I mean?  This is a good ordering (includes what I think are the only two board definitions that don't yet exist and will - attiny1634micr, and attiny861micr (I have hardware in hand, but haven't had a chance to build FW or test in the.... 5 months (yikes) since I got the boards fabbed and assembled a couple...).. Ordering in the core now is also fine, nobody seems to complain. The idea being to put the popular parts near the top, and the stuff nobody uses at the bottom 

attinyx41.name=ATtiny441/841 (No bootloader)
attinyx41opti.name=ATtiny441/841 (Optiboot)
attinyx41micr.name=ATtiny841 (Micronucleus / Wattuino)
attinyx5.name=ATtiny25/45/85 (No bootloader)
attinyx5opti.name=ATtiny45/85 (Optiboot)
attinyx5micr.name=ATtiny85 (Micronucleus / DigiSpark)
attinyx4.name=ATtiny24/44/84(a) (No bootloader)
attinyx4opti.name=ATtiny44/84(a) (Optiboot)
attinyx4micr.name=ATtiny84a (Micronucleus / California STEAM)
attiny1634.name=ATtiny1634 (No bootloader)
attiny1634opti.name=ATtiny1634 (Optiboot)
attiny1634micr.name=ATtiny1634 (Micronucleus / Azduino)
attinyx8.name=ATtiny48/88 (No bootloader)
attinyx8opti.name=ATtiny48/88 (Optiboot)
attinyx8micr.name=ATtiny88 (Micronucleus, MH-ET t88 w/16MHz CLOCK)
attinyx7.name=ATtiny87/167 (No bootloader)
attinyx7opti.name=ATtiny167/87 (Optiboot)
attinyx7micr.name=ATtiny167 (Micronucleus / DigiSpark Pro)
attinyx61.name=ATtiny261/461/861(a)
attinyx61opti.name=ATtiny461/861(a) (Optiboot)
attinyx61micr.name=ATtiny861(a) (Micronucleus / Azduino)
attiny828.name=ATtiny828 (No bootloader)
attiny828opti.name=ATtiny828 (Optiboot)
attinyx313.name=ATtiny2313(a)/4313 (No bootloader)
attiny43.name=ATtiny43 (No bootloader)

##########################################################################
#      boardname.name
##########################################################################

All properties. For ordering, I think ideal is:
all non-menu properties, alphabetical order
blank line
all boardname.menu.* properties.  The order of the menus here doesn't matter (though I think matching the above is desirable), but the orders of the options within the menu does, as that determines the order in which the IDE shows them Everything associated with a menu option grouped, options for each menu grouped. eg:
attinyx4.menu.clock.8internal=8 MHz (internal)
attinyx4.menu.clock.8internal.bootloader.low_fuses=0xE2
attinyx4.menu.clock.8internal.build.f_cpu=8000000L
attinyx4.menu.clock.8internal.build.clocksource=0
attinyx4.menu.clock.3internal=3 MHz (sorts above 8 internal, but we want it to show lower)
attinyx4.menu.clock.3internal.bootloader.low_fuses=0xE2
attinyx4.menu.clock.3internal.build.f_cpu=3000000L
attinyx4.menu.clock.3internal.build.clocksource=0

This is also fine, as long as it's consistent:

attinyx4.menu.clock.8internal=8 MHz (internal)
attinyx4.menu.clock.3internal=3 MHz (sorts above 8 internal, but we want it to show lower)
attinyx4.menu.clock.8internal.bootloader.low_fuses=0xE2
attinyx4.menu.clock.8internal.build.f_cpu=8000000L
attinyx4.menu.clock.8internal.build.clocksource=0
attinyx4.menu.clock.3internal.bootloader.low_fuses=0xE2
attinyx4.menu.clock.3internal.build.f_cpu=3000000L
attinyx4.menu.clock.3internal.build.clocksource=0

but not
attinyx4.menu.clock.3internal=3 MHz (sorts above 8 internal, but we want it to show lower)
attinyx4.menu.clock.8internal=8 MHz (internal)
attinyx4.menu.clock.8internal.bootloader.low_fuses=0xE2
attinyx4.menu.clock.8internal.build.f_cpu=8000000L
attinyx4.menu.clock.8internal.build.clocksource=0
attinyx4.menu.clock.3internal.bootloader.low_fuses=0xE2
attinyx4.menu.clock.3internal.build.f_cpu=3000000L
attinyx4.menu.clock.3internal.build.clocksource=0
```
It's to make it easier to read and update the file - so that I can check for missing/wrong entries. Being able to edit a json file and then urn if back into the boards.txt I think is the dream.

Thaaaaaanks! <3

**Comment by lerouxb** [comment]
You can start fixing some things :)

```
Failed tests:

  1) Format ATTinyCore/boards.txt:

      duplicate paths: Expected [ 'attiny1634.bootloader.extended_fuses',
  'attiny1634.bootloader.high_fuses',
  'attiny1634opti.bootloader.extended_fuses',
  'attiny1634opti.bootloader.high_fuses',
  'attinyx41micr.bootloader.extended_fuses',
  'attinyx41micr.bootloader.high_fuses',
  'attinyx41opti.bootloader.extended_fuses',
  'attinyx41opti.bootloader.high_fuses',
  'attinyx61opti.menu.clock.8internal.build.clocksource',
  'attinyx7.upload.maximum_data_size',
  'attinyx8opti.menu.clock.16external.build.clocksource' ] to be empty

      at /Users/lerouxb/src/arduino-config-parser/test/format.js:118:57

  2) Format ATTinyCore/externalprogrammers.txt:

      duplicate paths: Expected [ 'avrisp.program.tool' ] to be empty

      at /Users/lerouxb/src/arduino-config-parser/test/format.js:118:57

  3) Format ATTinyCore/platform.txt:

      duplicate paths: Expected [ 'recipe.hooks.savehex.presavehex.5.pattern.linux' ] to be empty

      at /Users/lerouxb/src/arduino-config-parser/test/format.js:118:57

  4) Format ATTinyCore/programmers.txt:

      duplicate paths: Expected [ 'avrisp.program.tool',
  'diamexusbisp.program.tool',
  'ponyser.program.tool',
  'usbtinyisp.program.extra_params' ] to be empty

      at /Users/lerouxb/src/arduino-config-parser/test/format.js:118:57

  18) Format DxCore/boards.txt:

      duplicate paths: Expected [ 'avrda.menu.startuptime.1',
  'avrdaopti.menu.startuptime.1',
  'avrdb.menu.startuptime.1',
  'avrdbopti.menu.chip.128DB28.bootloader.file',
  'avrdbopti.menu.chip.128DB28.bootloader.file',
  'avrdbopti.menu.chip.128DB28.bootloader.file',
  'avrdbopti.menu.chip.32DB28.bootloader.file',
  'avrdbopti.menu.chip.32DB28.bootloader.file',
  'avrdbopti.menu.chip.64DB28.bootloader.file',
  'avrdbopti.menu.chip.64DB28.bootloader.file',
  'avrdbopti.menu.chip.64DB28.bootloader.file',
  'avrdbopti.menu.startuptime.1' ] to be empty

      at /Users/lerouxb/src/arduino-config-parser/test/format.js:118:57

  21) Format megaTinyCore/boards.txt:

      duplicate paths: Expected [ 'atxy2.bootloader.tool', 'atxy2.menu.chip.402w.build.board' ] to be empty

      at /Users/lerouxb/src/arduino-config-parser/test/format.js:118:57
```

**Comment by SpenceKonde** [comment]
Told you there were bugs in the text files!

Most were non-scary/harmful but - non-28-pin AVR DB's didn't have bootloaders defined (it was only defined for the 28-pin one... once for every pin-count variant (at least this was a recent bug!). The two mentioning clocksource for attinycore were copy-paste followed by failure to change the clock option in  the pasted entry. In the case of the x61 opti, that took out *both* of the effected clock options. Both have been that way since the summer. 

Frustratingly I checked them into the wrong branch, heh. Will still be in next release though, just not merged into master before then.

**Reply by SpenceKonde** [reply]
All are also excellent examples of the kind of bug that the file format invites.

**Reply by SpenceKonde** [reply]
Oh, and also on the Dx-series parts, the 0ms SUT option wasn't actually available (no item in menu) because that first line that gives the name to show it as referenced the 1ms one

**Comment by lerouxb** [comment]
Here you go: https://github.com/lerouxb/arduino-config-parser

Hopefully enough to get you started. I think I'm done for today, but let me know if you'd like some more tools there. Perhaps a verification script? Something to convert it to another file format that might be more convenient for scanning for errors? Some code to easily manipulate the file?

**Comment by SpenceKonde** [comment]
My god getting other stuff out of the way so I could come back to ATTinyCore took fucking forever

Now I'm finally back here, pushing out ATTinyCore 1.5.0 to pick up the current changes in the repo to tide people over while I implement everything else that is planned. I think the next release is likely to be v2.0.0 - I need to make sure I nail all the questionably compatible changes at once (there are a bunch of awkward things that need to be fixed and could cause breakage - The PIN_xn defines were created before I realized Arduino official core used PIN_An for a different purpose - hence PIN_Pxn for port/bit notation to refer to pins - but the parts that already had PIN_xn defines kept them for backward compatibility. These are getting dropped. Bootloader upload speeds are getting changed (and in a somewhat counterintuitive way, too - 12 MHz boards will upload at 115200 via optiboot, but 16 MHz ones will be slower; classic AVRs don't do a great job of rendering 115200 baud from 16 MHz system clock it's off by about 2% and if that error is in the same direction as the RC oscillator's error (on the 841, 1634 and 828, it always is at Vcc > 4v); the new fractional baud rate generator is one of the many reasons that the new AVRs are better parts), pin defines across all the parts are being made consistent, which may mean awkward non-compliant ones that I don't think are in widespread use will be dropped, and so on) 

Now looking at your code and trying to figure out how to run it (haven't used Node before)

**Comment by per1234** [comment]
I'll share a couple things in case you find them useful:

There is an official Golang module for working with the Arduino "properties" data format used by boards.txt, platform.txt, programmers.txt, and library.properties:
https://github.com/arduino/go-properties-orderedmap
This is what the Arduino build system uses.
I see you already ran into the nightmare of dealing with the data structures:
https://github.com/lerouxb/arduino-config-parser#notes-about-the-json-structure
>The .txt format allows you to have something that is both an "object" (ie. it has properties) AND a simple value itself.

The way this is dealt with in Arduino Lint is to selectively, rather than programmatically, mapping the structure. What I mean is that sometimes a `.` indicates a level in the structure, but sometimes a `.` is just a character in the key name.

I mentioned Arduino Lint:
https://github.com/arduino/arduino-lint
This applies a bunch of rules to platform configuration files to check for specification compliance and best practices. Maybe you will find it useful in itself, or as a reference. There are also JSON schemas for these files:
https://github.com/arduino/arduino-lint/tree/main/etc/schemas

**Comment by SpenceKonde** [comment]
Thanks for your help on this, unfortunately I haven't been able to m,akle jheadway with JS - I can kind of trudge through python. but this noide stuiff is unfamiliar to be and I haven't been able to make use of it. 

Deal;ling with boards.txt is now thje blocking issue for fiurther progress on the core. I spent about 6  hours just trying to parse it and write it back out in python andjhave not been able ot pull it off/

**Reply by SpenceKonde** [reply]
This just isn't what I klnow how to work with. I know how hack at python I guess... but when I'm writing C  for embedded  with inline assembly, I can make an AVR walk on it's side. Buit these bigger computers are hard my brains address space is oml;ty 16 bit I thiink

**Comment by SpenceKonde** [comment]
What I would really like to get to is a dictionary in python,, from whidh I coudd dump various representations of it. 

Remember that in the case I care about (tjhe cores I maintain) each line is either:
1. A comment (# prefix): ignore. Comments will be put in later where appropriate. 
2. menu.something. Ignore, That can be handled trivially by hand,
3. boardname.,name=ATiny666 - this we keep. I'd propose soring the display names things as a __name property)
4. boardname.[build/upload/bootloader]*= Goes into the appropriate dictionary in the board (ex boards[boardname][build][core]="tiny
")
5. boardname.menu.option=value- Make a dictionaryt storing this nee menu as a dictionary, with __name=value. (eg boards[boardname][menu][option]={__name="valuie"}
6. boardname.menu.option.(stuff)=value would be added to that option's dictionarey, (eg boards[boardname][menu][option][(stuff)]=value

So then I could iterate over that and say, have it list which clock options each part has. What each part calls them (it's not even the same ffs,

and do the same for all the other options, so that I would know which boards cureently have which options, allowing me to determine where there are holes, becasue rthere are a great number of them



I am wondering if I should just attempt a generator script that given inputs, will generate a boards.txt file, and changing the result would just mean changing the arrays in the python scripts. I',m starting to lean towards this as the other options seem to be road blocks.

**Reply by per1234** [reply]
Here is a Python package for reading and writing the Arduino "properties" data format of the boards.txt, platform.txt, programmers.txt, and library.properties files:
https://github.com/jfjlaros/ordered-map
